validator_client/src/beacon_node_fallback.rs

```diff
@@ -613,80 +609,35 @@ impl<T: SlotClock, E: EthSpec> BeaconNodeFallback<T, E> {
-        // First pass: try `func` on all synced and ready candidates.
-        //
-        // This ensures that we always choose a synced node if it is available.
-        for candidate in &self.candidates {
-            match candidate.status(RequireSynced::Yes).await {
-                Err(CandidateError::NotSynced) if require_synced == false => {
-                    // This client is unsynced we will try it after trying all synced clients
-                    retry_unsynced.push(candidate);
-                }
-                Err(_) => {
-                    // This client was not ready on the first pass, we might try it again later.
-                    to_retry.push(candidate);
-                }
-                Ok(_) => try_func!(candidate),
-            }
+        // First pass: try `func` on all candidates. Candidate order has already been set in
+        // `update_all_candidates`. This ensures the most suitable node is always tried first.
+        let candidates = self.candidates.read().await;
+        for candidate in candidates.iter() {
+            try_func!(candidate);
         }
 
-        // Second pass: try `func` on ready unsynced candidates. This only runs if we permit
-        // unsynced candidates.
-        //
-        // Due to async race-conditions, it is possible that we will send a request to a candidate
-        // that has been set to an offline/unready status. This is acceptable.
-        if require_synced == false {
-            for candidate in retry_unsynced {
+        if !to_retry.is_empty() {
+            // Second pass. Some candidates did not return successfully. Try them again.
+            // Errors will still be shown for a node if it required a retry.
+            for candidate in to_retry.clone().iter() {
                 try_func!(candidate);
             }
         }
 
-        // Third pass: try again, attempting to make non-ready clients become ready.
-        for candidate in to_retry {
-            // If the candidate hasn't luckily transferred into the correct state in the meantime,
-            // force an update of the state.
-            let new_status = match candidate.status(require_synced).await {
-                Ok(()) => Ok(()),
-                Err(_) => {
-                    candidate
-                        .refresh_status(self.slot_clock.as_ref(), &self.spec, &self.log)
-                        .await
-                }
-            };
-
-            match new_status {
-                Ok(()) => try_func!(candidate),
-                Err(CandidateError::NotSynced) if require_synced == false => try_func!(candidate),
-                Err(e) => {
-                    results.push(Err((
-                        candidate.beacon_node.to_string(),
-                        Error::Unavailable(e),
-                    )));
-                }
-            }
-        }
-
         let errors: Vec<_> = results.into_iter().filter_map(|res| res.err()).collect();
 
         if !errors.is_empty() {
```

Comment:
Should we return `Ok` if all candidates return successfully after the second pass?